File: src/ltd/rust_lang/utils/queue/Deque.java
package ltd.rust_lang.utils.queue;

import ltd.rust_lang.utils.list.CircleDoublyLinkList;
import ltd.rust_lang.utils.list.MyList;

public class Deque<E> {
  private MyList<E> list = new CircleDoublyLinkList<>();

  public int size() {
    return list.size();
  }

  public boolean isEmpty() {
    return list.isEmpty();
  }

  public void enQueueRear(E element) {
    list.add(element);
  }

  public E deQueueFront() {
    return list.remove(0);
  }

  public void enQueueFront(E element) {
    list.insert(0, element);
  }

  public E deQueueRear() {
    return list.remove(list.size() - 1);
  }

  public E front() {
    return list.get(0);
  }

  public E rear() {
    return list.get(size() - 1);
  }

  public void clear() {
    list.clear();
  }
}

File: src/main/kotlin/ltd/rust_lang/utils/queue/Deque.kt
package ltd.rust_lang.utils.queue

import ltd.rust_lang.utils.list.LinkedList

class Deque<E> {
    private val list = LinkedList<E>()

    fun size(): Int {
        return list.size()
    }

    fun isEmpty(): Boolean {
        return list.isEmpty
    }

    fun offerFirst(e: E?): Boolean {
        return list.add(0, e)
    }

    fun offerLast(e: E?): Boolean {
        return list.add(e)
    }

    fun pollFirst(): E? {
        return list.remove(0)
    }

    fun pollLast(): E? {
        return list.remove(size() - 1)
    }

    fun peekFirst(): E? {
        return list[0]
    }

    fun peekLast(): E? {
        return list[size() - 1]
    }

    fun clear() {
        list.clear()
    }

    override fun toString(): String {
        return list.toString()
    }
}
